src/common/listener/Mapping.ts

```diff
@@ -0,0 +1,45 @@
+
+export type NeverMatch = null;
+
+/**
+ * Result of a Mapper. A result of the desired type if
+ * we match. undefined if we don't match.
+ * Return DoNotSetAnyGoals (null) to shortcut evaluation of the present set of rules,
+ * terminating evaluation and guarantee the return of undefined if we've reached this point.
+ * Only do so if you are sure
+ */
+export type MappingResult<V> = Promise<V | undefined | NeverMatch>;
+
+export type Mapper<F, V> = (from: F) => MappingResult<V>;
+
+/**
+ * Mapper from source to value, if it can be resolved.
+ * This is a central interface used throughout the SDM.
+ */
+export interface Mapping<F, V> {
+
+    /**
+     * Name of the PushMapping. Must be unique
+     */
+    readonly name: string;
+
+    /**
+     * Compute a value for the given event. Return undefined
+     * if we don't find a mapped value.
+     * Return DoNotSetAnyGoals (null) to shortcut evaluation of the present set of rules,
```

Comment:
this comment is not generic

which brings us to, what does NeverMatch mean?? There are two concerns in this mapping function: does it match, and what is the matched value.

Currently, "undefined" is overloaded to mean "not matched" while NeverMatch (null) is overloaded to mean "I matched! but with the value undefined!"

so it's now possible to match with undefined, but not with null.

very sneaky, hard to understand without a lot of reading ... what if we made a type that was SpecificMappingResult that had a value in it ... { kind: "PositiveMappingResult", value: V | undefined | null } so one that could contain any value, and is clear that it matches. Then returning V or returning "undefined" become shortcuts for this

Reply:
I fixed the comment. Changing to use a different return type has merit but it is non-trivial and I think it should go in a different PR, as these return semantics weren't introduced in this PR.